### ISSUE TITLE
IsGlobalProperty Property Function

### ISSUE THREAD

**Issue by TheRealPiotrP** [microsoft/msbuild#1692]
The ability to detect a property as being globally overridden enables SDKs to exert greater control over the behavior of multi-project or sln builds. A global property will influence the behavior of both referring projects and referenced projects. Unlike statically evaluated properties, global properties may have unintended consequences for referred projects. 

For example, consider RuntimeIdentifiers. If one tries to build a portable project while specifying a RuntimeIdentifier through a global override then the build should fail since restore would not have brought down appropriate assets. We should error out and help users fix their build setup. On the other hand, if a non-portable project references a portable project then users will build the non-portable project with a global override for Runtime Identifier. This global property will be passed along to the referenced project which will subsequently fail to build because it believes itself to be in a bad configuration.

By enabling detection of global properties, SDKs can provide their own policies around the passing of global properties to referenced projects. In the example above, if the referring project knows that it is being built with a global property that can put referenced projects in an unbuildable state then it can communicate to those projects that they can choose to ignore this global configuration. Without the ability to detect global properties, the referenced project must assume that it was misconfigured and therefore fail to build.

There is a workaround to this feature ask, currently in PR to the .NET SDK:
``` 
    <!-- We need to determine if the RuntimeIdentifier was globally set. This check takes advantage of the fact that global properties cannot be overwritten. -->
    <PropertyGroup>
      <!-- First, backup the value of the RuntimeIdentifier -->
      <GlobalRuntimeIdentifier>$(RuntimeIdentifier)</GlobalRuntimeIdentifier>
      <!-- Now set the RuntimeIdentifier to a bogus value. If it is global, this change will be ignored. -->
      <RuntimeIdentifier>__________</RuntimeIdentifier>
    </PropertyGroup>

    <!-- If the RuntimeIdentifier was not set globally, put back its local value. -->
    <PropertyGroup Condition=" '$(RuntimeIdentifier)' != '$(GlobalRuntimeIdentifier)'">
      <RuntimeIdentifier>$(GlobalRuntimeIdentifier)</RuntimeIdentifier>
      <!-- Null out the GlobalRuntimeIdentifier since RuntimeIdentifier is not set globally. -->
      <GlobalRuntimeIdentifier></GlobalRuntimeIdentifier>
    </PropertyGroup>
  
    <!-- If the RuntimeIdentifier was set globally, pass it to GetTargetFrameworkProperties. -->
    <PropertyGroup Condition=" '$(GlobalRuntimeIdentifier)' != '' ">  
  <AdditionalPropertiesForGetTargetFrameworkProperties>;ReferringRuntimeIdentifier=$(GlobalRuntimeIdentifier)</AdditionalPropertiesForGetTargetFrameworkProperties>
    </PropertyGroup>
```

However, as can be seen above, the workaround requires quite a bit of code and depends on an implementation detail of global properties that is not obvious to most.

**Comment by rainersigwald** [comment]
Notes from looking into this a few days ago:

>looks like it's not trivial to build a property function for `IsGlobalProperty`. The [part of the engine that evaluates property functions](https://github.com/Microsoft/msbuild/blob/487db4a80806347f2f5ec8e41c7cf428dab06c92/src/Build/Evaluation/Expander.cs#L3006) doesn't have a direct link back to the project, only an `IPropertyProvider` interface that doesn't distinguish between global and local properties. It's still possible but not super easy

The engine could either pass something down to that level that has a link back to something that knows about global properties, or [`IPropertyProvider`](https://github.com/Microsoft/msbuild/blob/487db4a80806347f2f5ec8e41c7cf428dab06c92/src/Build/Evaluation/IPropertyProvider.cs) could be augmented to have a `bool IsGlobal` or similar.

I'm not willing to do either for v15.1 RTW, but I don't have a moral objection to either in the longer term. Since this is in core property-evaluation code, an implementation would need to be careful about the memory and compute performance impact of carrying the extra data around.

**Comment by rainersigwald** [comment]
This can now be implemented as a custom task using #4925, so I'm going to close this.
